momaland/envs/ingenious/ingenious_base.py

```diff
@@ -0,0 +1,427 @@
+            self.tiles_bag = int(NUM_TILES / len(diff_color_combinations + same_color_combinations)) * (
+                diff_color_combinations + same_color_combinations
+            )
+        # Shuffle the tiles bag
+        self.random.shuffle(self.tiles_bag)
+
+    def set_action_index(self, index):
+        """Apply the corresponding action for the given index on the board."""
+        """If selected actions is not a legal move, return False"""
+        if self.masked_action[index] == 0:
+            assert "illegal move:choose a masked action(0)"
+            return False
+        if (index not in self.first_round_pos) and self.first_round:
+            assert "illegal move: choose a action legally in first round"
+            return False
+        """Hex Coordinate: h1,h2 ;  Tile to play: card"""
+        h1, h2, card = self.action_index_map[index]
+        agent_i = self.agent_selector
+        agent = self.agents[agent_i]
+        if card >= len(self.p_tiles[agent]):
+            assert "illegal move: choosing tile out of hand(happening after ingenious)"
+            return False
+        """Extract the certain tile (color1 , color2) as (c1,c2)"""
+        c1, c2 = self.p_tiles[agent][card]
+        # Translate Hex Coordinate to Offset Coordinate(x,y)
+        x1, y1 = Hex2ArrayLocation(h1, self.board_size)
+        x2, y2 = Hex2ArrayLocation(h2, self.board_size)
+        flag = False
+        for item in self.p_tiles[agent]:
+            # print(item)
+            if (c1, c2) == item:
+                self.p_tiles[agent].remove(item)
+                flag = True
+                break
+            if (c2, c1) == item:
+                self.p_tiles[agent].remove(item)
+                flag = True
+                break
+        if not flag:
+            assert "illegal move: set the tile to the coordinate unsuccessfully"
+            return False
+        """Update the mask_action list after the action"""
+        self.legal_move.remove(index)
+        self.board_array[x1][y1] = c1
+        self.board_array[x2][y2] = c2
+        self.exclude_action(h1)
+        self.exclude_action(h2)
+        skip_flag = False
+        """Update score through checking 5 neighboring directions for h1 and h2 independently"""
+        point = 0
+        for i in range(0, 6):
+            h1n = hex_neighbor(h1, i)
+            if h1n == h2:
+                continue
+            # print(h1n)
+            while h1n in self.board_hex:
+                x, y = Hex2ArrayLocation(h1n, self.board_size)
+                # print(x,y,h1n)
+                if self.board_array[x][y] == c1:
+                    point += 1
+                else:
+                    break
+                h1n = hex_neighbor(h1n, i)
+                # print(h1n)
+        self.score[agent][c1] += point
+        if self.score[agent][c1] > self.limitation_score:
+            skip_flag = True
+            self.score[agent][c1] = 0
```

Comment:
As I understand, "Bonus Move" with calling "ingenious" is a general rule not just in solo game or partnerships(why the game is called 'ingenious'). I implement the game in the competitive multi-agent setting. Single agent situation is not considered in the case for that the motivation of MOMALAND is multi-agent morl.

Reply:
I agree with the fact that the bonus move is always played (regardless of the setting), I was referring more to the fact that the score of the specific color is reset here during "ingenious". As far as I was able to gather from the rules I think the resetting only happens in the single-player of the game. In the normal multi-agent setting, I think that you just keep the maximum score of the color (in this case 18) once you reach ingenious.

_"Your counters cannot advance beyond the final spaces (value 18) on the score board."_

Otherwise the rule which states that a player wins the game in the rare case that a player has reached the end of all colors also wouldn't be possible (as the colors would have reset in this case).

_"In the very rare case that one player succeeds in advancing all 6 counters to the end of his score board, he wins instantly."_

Reply:
Yes, you are right. My mistake. In multi agent competitive case, maybe it should just restrain the score at the maximal credict. Like these:
      if self.score[agent][c1] > self.limitation_score:
            skip_flag = True
            self.score[agent][c1] = self.limitation_score
        if self.score[agent][c2] > self.limitation_score:
            skip_flag = True
            self.score[agent][c2] = self.limitation_score

Reply:
Fixed.